### PR TITLE
Simplify serialization.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 1]
```diff
@@ -576,17 +576,12 @@ public Type ReadType()
...
-            _reader.ReadByte();
-            return _binderSnapshot.GetTypeReaderFromId(this.ReadInt32());
-        }
-
         private object ReadObject()
         {
             var id = _objectReferenceMap.GetNextReferenceId();
 
-            var typeReader = this.ReadTypeReader();
+            _reader.ReadByte();
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: ❕ Please add a comment so it's clear what this byte is.

❔ This appears to be the `EncodingKind.Object` value written in `WriteObjectWorker`. Is it used elsewhere? If not we might want to remove it.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

File: src/Compilers/Core/Portable/Serialization/ObjectReader.cs [hunk 2]
```diff
@@ -32,16 +30,18 @@ internal sealed partial class ObjectReader : IDisposable
         /// this version, just change VersionByte2.
         /// </summary>
         internal const byte VersionByte1 = 0b10101010;
-        internal const byte VersionByte2 = 0b00000111;
+        internal const byte VersionByte2 = 0b00001000;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Serialization/ObjectReader.cs]: Changed hte serialization formt for how we write out IObjectWritables.  Previously it was "ObjectTag-TypeTag-TypeId-...data...".  Now it is "ObjectTag-TypeId-...data...".

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz .  Continued squeezing of perf for stuff that's showing up while i profile OOP.  

Serialization is a huge perf area for us.  We're constantly dumping out syntax trees to memory mapped files, and having to read them in.  There's a ton of recursion here along with reading/writing millions of nodes.  So every bit we can squeeze helps.

**Comment by MattGertz** [comment]
Approved pending tests.

**Pull request merged**